src/core/server/elasticsearch/status.test.ts

```diff
@@ -0,0 +1,218 @@
+      meta: {
+        warningNodes: [nodeInfo],
+        incompatibleNodes: [nodeInfo],
+      },
+    });
+  });
+
+  it('emits status updates when node compatibility changes', () => {
+    const nodeCompat$ = new Subject<NodesVersionCompatibility>();
+
+    const statusUpdates: ServiceStatus[] = [];
+    const subscription = calculateStatus$(nodeCompat$).subscribe(status =>
+      statusUpdates.push(status)
+    );
+
+    nodeCompat$.next({
+      isCompatible: false,
+      kibanaVersion: '2.1.1',
+      incompatibleNodes: [],
+      warningNodes: [],
+      message: 'Unable to retrieve version info',
+    });
+    nodeCompat$.next({
+      isCompatible: false,
+      kibanaVersion: '2.1.1',
+      incompatibleNodes: [nodeInfo],
+      warningNodes: [],
+      message: 'Incompatible with Elasticsearch',
+    });
+    nodeCompat$.next({
+      isCompatible: true,
+      kibanaVersion: '1.1.1',
+      warningNodes: [nodeInfo],
+      incompatibleNodes: [],
+      message: 'Some nodes are incompatible',
+    });
+    nodeCompat$.next({
+      isCompatible: true,
+      kibanaVersion: '1.1.1',
+      warningNodes: [],
+      incompatibleNodes: [],
+    });
+
+    expect(statusUpdates.map(({ level, summary }) => ({ level, summary }))).toMatchInlineSnapshot(`
+      Array [
+        Object {
+          "level": 2,
+          "summary": "Waiting for Elasticsearch",
+        },
+        Object {
+          "level": 3,
+          "summary": "Unable to retrieve version info",
+        },
+        Object {
+          "level": 3,
+          "summary": "Incompatible with Elasticsearch",
+        },
+        Object {
+          "level": 1,
+          "summary": "Some nodes are incompatible",
+        },
+        Object {
+          "level": 0,
+          "summary": "Elasticsearch is available",
+        },
+      ]
+    `);
+
+    subscription.unsubscribe();
+  });
+
+  it('does not emit duplicate status updates', () => {
+    const nodeCompat$ = new Subject<NodesVersionCompatibility>();
+
+    const statusUpdates: ServiceStatus[] = [];
+    const subscription = calculateStatus$(nodeCompat$).subscribe(status =>
+      statusUpdates.push(status)
+    );
+
+    nodeCompat$.next({
+      isCompatible: false,
+      kibanaVersion: '2.1.1',
+      incompatibleNodes: [nodeInfo],
+      warningNodes: [],
+      message: 'Incompatible with Elasticsearch',
+    });
+    nodeCompat$.next({
+      isCompatible: false,
+      kibanaVersion: '2.1.1',
+      incompatibleNodes: [nodeInfo],
+      warningNodes: [],
+      message: 'Incompatible with Elasticsearch',
+    });
+    nodeCompat$.next({
+      isCompatible: false,
+      kibanaVersion: '2.1.1',
+      incompatibleNodes: [nodeInfo],
+      warningNodes: [],
+      message: 'Incompatible with Elasticsearch',
+    });
+
+    expect(statusUpdates.map(({ level, summary }) => ({ level, summary }))).toMatchInlineSnapshot(`
+      Array [
+        Object {
+          "level": 2,
```

Comment:
Yeah I think the sorting property that I liked from a regular integer enum can be expressed as a utility. String readability is probably more important. 👍

Reply:
Actually on second thought, I think we're more often concerned with comparisons than we are with the string output (which is really only needed in the API outputs).

If we move to string enums, we would probably want to include a utility like this:
```ts
export const statusComparator = {
  gt: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === 1,
  gte: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) >= 0,
  lt: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === -1,
  lte: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) <= -1,
  eq: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === 0,
  compareFn,
};
```

However, I think this will be far less discoverable. Since doing comparisons like `esStatus >= unavailable` may be very common, I don't think this is an improvement. WDYT?

Reply:
> However, I think this will be far less discoverable. Since doing comparisons like esStatus >= unavailable may be very common, I don't think this is an improvement. WDYT?

Optional: We can define each level as an object or a class instance with a `valueOf` method to make them comparable.
```js
const available = {
  toString() {
    return 'available';
  },
  valueOf() {
    return 0;
  },
};

const degraded = {
  toString() {
    return 'degraded';
  },
  valueOf() {
    return 1;
  },
};
const levels = {
  available,
  degraded,
};

console.log(levels.available === levels.available);
console.log(levels.available > levels.degraded);
console.log(levels.available === levels.degraded);
console.log(levels.available < levels.degraded);
```

Reply:
```ts
const available = {
  toString() {
    return 'available';
  },
  valueOf() {
    return 0;
  },
};
```

Would love to have java `enum` class implementation in TS. that would really solves this kind of issues in a graceful way...

Reply:
Good idea, switched to this pattern. It doesn't work for array sorting, but I'm less concerned about that being a common use case.